source/modulo_controllers/src/BaseControllerInterface.cpp

```diff
@@ -539,6 +542,59 @@ void BaseControllerInterface::add_service(
   }
 }
 
+void BaseControllerInterface::add_tf_listener() {
+  if (this->get_node() == nullptr) {
+    throw modulo_core::exceptions::CoreException("Failed to add TF buffer and listener: Node is not initialized yet.");
+  }
+
+  if (this->tf_buffer_ == nullptr || this->tf_listener_ == nullptr) {
+    RCLCPP_DEBUG(this->get_node()->get_logger(), "Adding TF buffer and listener.");
+    console_bridge::setLogLevel(console_bridge::CONSOLE_BRIDGE_LOG_NONE);
+    this->tf_buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_node()->get_clock());
+    this->tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*this->tf_buffer_);
+  } else {
+    RCLCPP_DEBUG(this->get_node()->get_logger(), "TF buffer and listener already exist.");
+  }
+}
+
+state_representation::CartesianPose BaseControllerInterface::lookup_transform(
+    const std::string& frame, const std::string& reference_frame, const tf2::TimePoint& time_point,
+    const tf2::Duration& duration) {
+  auto transform = this->lookup_ros_transform(frame, reference_frame, time_point, duration);
+  state_representation::CartesianPose result(frame, reference_frame);
+  translators::read_message(result, transform);
+  return result;
+}
+
+state_representation::CartesianPose BaseControllerInterface::lookup_transform(
+    const std::string& frame, const std::string& reference_frame, double validity_period,
+    const tf2::Duration& duration) {
+  auto transform =
+      this->lookup_ros_transform(frame, reference_frame, tf2::TimePoint(std::chrono::microseconds(0)), duration);
+  if (validity_period > 0.0
+      && (this->get_node()->get_clock()->now() - transform.header.stamp).seconds() > validity_period) {
+    throw modulo_core::exceptions::LookupTransformException("Failed to lookup transform: Latest transform is too old!");
+  }
+  state_representation::CartesianPose result(frame, reference_frame);
+  translators::read_message(result, transform);
+  return result;
+}
+
+geometry_msgs::msg::TransformStamped BaseControllerInterface::lookup_ros_transform(
+    const std::string& frame, const std::string& reference_frame, const tf2::TimePoint& time_point,
+    const tf2::Duration& duration) {
+  if (this->tf_buffer_ == nullptr || this->tf_listener_ == nullptr) {
+    throw modulo_core::exceptions::LookupTransformException(
+        "Failed to lookup transform: To TF buffer / listener configured.");
```

Comment:
```suggestion
        "Failed to lookup transform: No TF buffer / listener configured.");
```

Reply:
You can fix that in the Component interface too please :smile:

Reply:
ah nice, didn't even notice..